src/compiler/checker.ts

```diff
@@ -12100,6 +12100,20 @@ module ts {
             return false;
         }
 
+        function isValidTypeOfIndexSignatureParameter(parameter: ParameterDeclaration): boolean {
+            if (parameter.type.kind === SyntaxKind.StringKeyword || parameter.type.kind === SyntaxKind.NumberKeyword) {

```

Comment:
I would not do this check.

Instead, just make the call to getTypeFromSomething(parameter.type), and check if it is the string, number, or numberlike type.

Note: parameter.type may be null iirc.